src/main/java/org/testng/internal/ConfigurationGroupMethods.java

```diff
@@ -120,27 +128,36 @@ private void log(String string) {
     Utils.log("ConfigurationGroupMethods", 2, string);
   }
 
-  public Map<String, List<ITestNGMethod>> getBeforeGroupsMap() {
-    return m_beforeGroupsMethods;
+  public List<ITestNGMethod> getBeforeGroupMethodsForGroup(String group) {
+    synchronized (beforeGroupsThatHaveAlreadyRun) {
+      return retrieve(beforeGroupsThatHaveAlreadyRun, m_beforeGroupsMethods, group);
+    }
   }
 
-  public Map<String, List<ITestNGMethod>> getAfterGroupsMap() {
-    return m_afterGroupsMethods;
+  public List<ITestNGMethod> getAfterGroupMethodsForGroup(String group) {
+    synchronized (afterGroupsThatHaveAlreadyRun) {
+      return retrieve(afterGroupsThatHaveAlreadyRun, m_afterGroupsMethods, group);
+    }
   }
 
   public void removeBeforeGroups(String[] groups) {
     for(String group : groups) {
-//      log("Removing before group " + group);
       m_beforeGroupsMethods.remove(group);
     }
   }
 
   public void removeAfterGroups(Collection<String> groups) {
     for(String group : groups) {
-//      log("Removing before group " + group);
       m_afterGroupsMethods.remove(group);
     }
+  }
 
+  private static List<ITestNGMethod> retrieve(Set<String> tracker, Map<String, List<ITestNGMethod>> map, String group) {
+    if (tracker.contains(group)) {
+      return new ArrayList<>();

```

Comment:
Maybe emptyList? I don't see why the list should be modified outside.

Reply:
Fixed.